gk/fib.c

```diff
@@ -1533,3 +1534,295 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+				sizeof(dentry->grantor_ip)) ==
+				NULL) {
+			RTE_LOG(ERR, GATEKEEPER, "gk: %s: failed to convert a number to an IPv6 address (%s)\n",
+				__func__, strerror(errno));
+			rte_free(eth_tbl);
+			return -1;
+		}
+	} else {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"Invalid IP protocol (%hhu) at %s with lcore %u\n",
+			fib->u.grantor.gt_addr.proto,
+			__func__, rte_lcore_id());
+		rte_free(eth_tbl);
+		return -1;
+	}
+
+	ret = fillup_gk_fib_ether_entry(fib->u.grantor.eth_cache, eth_tbl);
+	if (ret < 0) {
+		rte_free(eth_tbl);
+		return -1;
+	}
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gateway_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret;
+	struct gk_fib_ether_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	ret = fillup_gk_fib_ether_entry(fib->u.gateway.eth_cache, eth_tbl);
+	if (ret < 0) {
+		rte_free(eth_tbl);
+		return -1;
+	}
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_neighbor_fib_dump_entry(uint16_t proto,
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int num_entries = 0;
+	uint32_t next = 0;
+	int32_t index;
+	const void *key;
+	void *data;
+	struct neighbor_hash_table *neigh_ht;
+	struct gk_fib_ether_entry *eth_tbl;
+	struct gk_fib_ether_entry *entry;
+
+	if (proto == ETHER_TYPE_IPv4)
+		neigh_ht = &fib->u.neigh;
+	else
+		neigh_ht = &fib->u.neigh6;
+
+	eth_tbl = rte_calloc(NULL, neigh_ht->tbl_size,
+		sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		int ret;
+		struct ether_cache *eth_cache = &neigh_ht->cache_tbl[index];
+		entry = &eth_tbl[num_entries++];
+
+		ret = fillup_gk_fib_ether_entry(eth_cache, entry);
+		if (ret < 0) {
+			rte_free(eth_tbl);
+			return -1;
+		}
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	dentry->num_ether_entries = num_entries;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gk_fib_dump_entry(uint16_t proto,
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(
+			proto, dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+struct gk_fib_dump_entry *
```

Comment:
I think we should add a comment that says callers of this function need to deallocate the `struct gk_fib_dump_entry *` that is returned using `rte_free()`.